geopolars/src/spatial_index.rs

```diff
@@ -0,0 +1,437 @@
+use geo::{
+    prelude::BoundingRect, Geometry, Line, LineString, MultiLineString, MultiPoint, MultiPolygon,
+    Point, Polygon,
+};
+use polars::prelude::{DataFrame, JoinType, NamedFrom, PolarsError, Result, Series};
+use rstar::{RTree, RTreeObject, AABB};
+
+use crate::util::{geom_at_index, iter_geom, Predicate};
+
+pub struct SpatialJoinArgs<'a> {
+    pub join_type: JoinType,
+    pub predicate: Predicate,
+    pub l_suffix: &'a str,
+    pub r_suffix: &'a str,
+    pub l_index: Option<&'a SpatialIndex>,
+    pub r_index: Option<&'a SpatialIndex>,
+}
+
+impl<'a> Default for SpatialJoinArgs<'a> {
+    fn default() -> Self {
+        Self {
+            join_type: JoinType::Inner,
+            predicate: Predicate::Intersects,
+            l_suffix: "_left",
+            r_suffix: "_right",
+            l_index: None,
+            r_index: None,
+        }
+    }
+}
+
+pub fn spatial_join(
+    lhs: &DataFrame,
+    rhs: &DataFrame,
+    options: SpatialJoinArgs,
+) -> Result<DataFrame> {
+    use geo::algorithm::{contains::Contains, intersects::Intersects};
+
+    let lhs_geometry = lhs.column("geometry")?;
+    let rhs_geometry = rhs.column("geometry")?;
+
+    let spatial_index_left: SpatialIndex = lhs_geometry.try_into().map_err(|_| {
+        PolarsError::ComputeError(std::borrow::Cow::Borrowed(
+            "Failed to generate the spatial index for the left dataframe",
+        ))
+    })?;
+
+    let spatial_index_right: SpatialIndex = rhs_geometry.try_into().map_err(|_| {
+        PolarsError::ComputeError(std::borrow::Cow::Borrowed(
+            "Failed to generate the spatial index for the right dataframe",
+        ))
+    })?;
```

Comment:
Not 100% sure how to optionally construct the index if none is passed. I tried 

```rust 
let spatial_index_left: &SpatialIndex = if let Some(index_ref) = options.l_index{
    index_ref
}
else{
    let new_index = lhs_geometry.try_into().map_err(|_| {
        PolarsError::ComputeError(std::borrow::Cow::Borrowed(
            "Failed to generate the spatial index for the left dataframe",
        ))?;
   &new_index
}
```

but this fails because new_index obviously goes out of scope at the end of the if let statement.

Not had to do this in rust before so not 100% sure of the idiomatic rust way of doing this. It seems like it should be possible to tell rust to hold on to new_index till the end of the function, but I am not 100% sure how to implement it (guessing some sort of lifetime?) @kylebarron thoughts?

Reply:
Made these Arcs which helps resolve that problem and also makes it easier to share indexes in the future between threads etc.

Reply:
> because new_index obviously goes out of scope at the end of the if let statement

Hmm. So the issue here is we want to handle both when we don't own the memory (spatial index is passed in) or when we do own the memory (we generate it)? I don't know what the idiomatic rust way of doing that would be.

> Made these Arcs which helps resolve that problem

That seems fine to me. We can always optimize further in the future if needed